### PR TITLE
New and improved `FixedDecimal` type parser

### PR DESCRIPTION
**Pull request by Drvi**
This is a new implementation which differs from the previous one in that it:
* more idiomatically uses `Parsers.jl` abstractions and integrates better with its components. Before I was manually detecting ends of fields instead of bailing on the first invalid byte... and as a result, we weren't correctly stripping trailing whitespace and would just call the value invalid instead, which was too strict.
* should be a bit faster; we still do more than one pass through the data so there is still some performance left on the table, but we're doing it less wastefully now -- before we'd make the initial scan to try to validate the overall structure, find positions of decimal point, exponent. But, we'd also jump around to try and skip leading/trailing zeroes. Now there only ever should be two passes.
* has a more consistent rounding behavior for `RoundNearest`, consider parsing a `FixedDecimal{Int,0}` from: 
```
in   master  PR `round(parse(Float64, in))`
0.5       0   0        0.0
0.51      0   1        1.0
1.5       2   2        2.0
1.51      2   2        2.0
```
While `master` was trying to be as faithful as possible with the original implementation from FixedPointDecimals.jl, I think we should try to be more consistent with how Base rounding works.

We were also not strict enough with `RoundThrows` as we originally skipped the rounding check when we knew all of the parsed digits were past the precision of the decimal (we never really exposed this parsing mode through ChunkedCSV though).
* supports parsing from IO streams, not just byte buffers
* has much-improved test coverage

I think that after we review here, we should move this to `FixedPointDecimals` and rebase the `Base.parse` on top of results from this type parser. We could also get rid of `_FixedDecimal` here and use `FixedDecimal` directly.

### CODE REVIEW

File: src/fixed_decimals_utils.jl [hunk 1]
```diff
@@ -127,284 +131,304 @@ _typemaxbytes(::Type{Int128}, i, is_neg) = @inbounds NTuple{39,UInt8}((0x31, 0x3
...
+            elseif UInt8('1') <= b <= UInt8('9')
+                !found_nonzero && (first_byte_to_parse = pos)
+                found_nonzero = true
+                last_nonzero_digit_pos = pos
+                pos += 1
+                Parsers.incr!(source)
+                num_digits += 1
+                Parsers.eof(source, pos, len) && (code |= Parsers.EOF; @goto done_scanning) # Input is something like "11<EOF>"
+            elseif has_groupmark && b == groupmark
+                # More than one groupmark in a row is invalid
```
Comment (nickrobinson251) [inline src/fixed_decimals_utils.jl:200]: was confused by "in a row" for a second there (thinking about e.g. a row in a CSV file)
```suggestion
                # More than one groupmark consecutively is invalid
```

File: src/fixed_decimals_utils.jl [hunk 2]
```diff
@@ -127,284 +131,304 @@ _typemaxbytes(::Type{Int128}, i, is_neg) = @inbounds NTuple{39,UInt8}((0x31, 0x3
...
+                prev_b == groupmark && @goto invalid # Input is something like "10,,"
+                pos += 1
+                Parsers.incr!(source)
+                found_nonzero && (ngroupmarks += 1)
+                Parsers.eof(source, pos, len) && @goto eof_invalid # Input is something like "10,<EOF>"
+            else
+                break
```
Comment (nickrobinson251) [inline src/fixed_decimals_utils.jl:207]: can this `else; break` clause have a comment? something like
```
else # we've hit a decimal
    break
```
actually, is that the only case we expect? also exponent `e` or `E`? should we `@assert` this?

File: src/fixed_decimals_utils.jl [hunk 3]
```diff
@@ -127,284 +131,304 @@ _typemaxbytes(::Type{Int128}, i, is_neg) = @inbounds NTuple{39,UInt8}((0x31, 0x3
...
+        !Parsers.ok(exp_code) && @goto invalid  # Input is something like "1ea"
+        fractional_part_length -= exp_val
     end
 
-    decimal_shift = f - number_of_fractional_digits
+    @label done_scanning
+    !found_digit && @goto invalid # Input is something like "a"
+    (T <: Unsigned && is_neg && found_nonzero) && @goto invalid # Output is something like "FixedDecimal{UInt64,8}(-123)"
+
+    # At this point we have visited all the relevant bytes, now we need to decide what do next
```
Comment (nickrobinson251) [inline src/fixed_decimals_utils.jl:280]: can everything below here be moved to its own function, where we pass in this info? if that's viable, i think it'd be a good idea, as i think it'd make it much easier to chase down any issues in future if there was a split in the code where we could easily get out / put in the various bits of info (fractional_part_length, num_digits, ...)

File: test/decimals.jl [hunk 4]
```diff
@@ -1,225 +1,337 @@
...
+           ("00001", 0),
+           ("000" * ("0" ^ 99) * "1", 0),
+           ("009", 0),
+           ("009" * ("9" ^ 99), 0),
+           ("01001", 1),
+           ("010" * ("0" ^ 99) * "1", 1),
+           ("019", 1),
+           ("019" * ("9" ^ 99), 1),
+        ]
+            test_decimal_parser(T, 1, insertchar(base_str, 2, "."); mode, expect=expected_value)
```
Comment (nickrobinson251) [inline test/decimals.jl:206]: i find all the tests in this file that are building up strings and using insertchar _very_ hard to follow and see that they're correct

can we instead just write out the actual numbers in full?

or, if that's too unwieldy, maybe put then in a plain text file along with the expected value and read them in

ideally it'd be easy to read off `rounding_mode, decimal_number, expected_value` all in one place and see that the test cases are correct

i think this is important because for the kind of subtle, manually unrolled code in this PR, we're relying on having excellent tests

File: test/decimals.jl [hunk 5]
```diff
@@ -1,225 +1,337 @@
...
+        @test inexact == _inexact_io
+        ok && (@test pos == (length(buf) + (UInt8(last(buf)) != UInt8(DELIM))))
+        ok && (@test pos_io == (length(buf) + (UInt8(last(buf)) != UInt8(DELIM))))
+        ok && (@test Parsers.eof(code))
+        ok && (@test Parsers.eof(code_io))
+        isnothing(expect) || (@test val == expect)
+        isnothing(expect) || (@test val_io == expect)
+        if mode == RoundNearest
+            res = Parsers.xparse(ChunkedCSV._FixedDecimal{T,f}, buf, 1, length(buf), _OPTIONS)
+            @test Parsers.eof(res.code) == (isempty(buf) || (UInt8(last(buf)) != UInt8(DELIM)))
```
Comment (nickrobinson251) [inline test/decimals.jl:36]: please can you add comments saying what each of these is testing and why?

as is, i don't follow why these are right/important things to test or if something is missing

**Review by nickrobinson251 — COMMENTED** [review]
Please can you make the test code easier to follow, so the tests are more _obviously correct_? I don't feel comfortable approving while i easily follow the tests

It looks like you've got really good coverage there, so i think it's just a matter of making things more explicit / a little less terse

File: test/decimals.jl [hunk 6]
```diff
@@ -1,225 +1,337 @@
...
+    macro testctx(ex)
+        esc(ex)
+    end
+end
+
+const DELIM=';';
+const GROUP=',';
+const DECIM='.';
+const _OPTIONS = Parsers.Options(groupmark=GROUP, decimal=DECIM, delim=DELIM, quoted=true)
+function test_decimal_parser(::Type{T}, f, buf::AbstractString; ok=true, inexact=false, mode=RoundNearest, expect=nothing) where {T}
```
Comment (nickrobinson251) [inline test/decimals.jl:19]: what is `f`? i think this function could use some documentation

File: test/decimals.jl [hunk 7]
```diff
@@ -1,225 +1,337 @@
...
+else
+    macro testctx(ex)
+        esc(ex)
+    end
+end
+
+const DELIM=';';
+const GROUP=',';
+const DECIM='.';
+const _OPTIONS = Parsers.Options(groupmark=GROUP, decimal=DECIM, delim=DELIM, quoted=true)
```
Comment (nickrobinson251) [inline test/decimals.jl:18]: do we need to test any other options? (if not, why not?)

File: src/fixed_decimals_utils.jl (hunk 2 above)
Comment (Drvi) [reply src/fixed_decimals_utils.jl:207]: This just means we're done with the "integer" part, what can follow is a decimal, E/e, quote, newline, delim, space... From what gather from other type parsers, the idiomatic usage is to parse up all valid bytes or end on the first invalid byte and leave the other components to handle recovery from that (e.g. skipping the end of the field and setting the relevant `code`s). I'll add a comment, if that is helpful.

File: test/decimals.jl (hunk 4 above)
Comment (Drvi) [reply test/decimals.jl:206]: Since we often unroll for each valid precision of each supported integer type would mean adding up to 153 strings per input string, per test. I don't think people are going to read thousands of lines of inputs just check the tests are correct manually, ideally, they'd only need to check how the test cases were constructed... which is what we have in the file. It is exactly the because of how subtle the code is that I chose exhaustive testing which sadly doesn't scale as far as reading each test case goes. Are there any more scalable approaches that would make the tests seem more credible?

File: test/decimals.jl (hunk 7 above)
Comment (Drvi) [reply test/decimals.jl:18]: I'll add a testset which tries a bunch of different options. :+1:

File: test/decimals.jl (hunk 4 above)
Comment (Drvi) [reply test/decimals.jl:206]: Would adding some examples in form of a comment help?

Comment (Drvi) [reply test/decimals.jl:206]: Or is it only about the tests where I do the insertchar thing?

File: src/fixed_decimals_utils.jl (hunk 2 above)
Comment (nickrobinson251) [reply src/fixed_decimals_utils.jl:207]: thanks, makes sense -- comment sounds good

File: test/decimals.jl (hunk 4 above)
Comment (nickrobinson251) [reply test/decimals.jl:206]: yeah, it's the `insertchar` tests i can't easily follow

Comment (nickrobinson251) [reply test/decimals.jl:230]: yeah, they help

**Review by nickrobinson251 — APPROVED** [review]

**Comment by NHDaly** [comment]
> > Where would be a good place to document that? maybe as a comment on the typeparser?
> 
> I think the rounding during parsing in FixedPointDecimals is just wrong, it appears to be inconsistent with regular rounding; on FixedPointDecimals master:
> 
> ```julia
> julia> reinterpret(FixedDecimal{Int,2}, 51)
> FixedDecimal{Int64,2}(0.51)
> 
> julia> round(reinterpret(FixedDecimal{Int,2}, 51)) # round to integer
> FixedDecimal{Int64,2}(1.00)
> 
> julia> parse(FixedDecimal{Int,0}, "0.51") # also rounds to integer because we don't allow any fractional digits
> FixedDecimal{Int64,0}(0)
> ```

ooh, bummer. Can you file an issue in the FixedDecimals repo about that? We should really be consistent between these two methods of parsing 👍

File: src/fixed_decimals_utils.jl [hunk 8]
```diff
@@ -127,284 +131,381 @@ _typemaxbytes(::Type{Int128}, i, is_neg) = @inbounds NTuple{39,UInt8}((0x31, 0x3
 _typemaxbytes(::Type{UInt8}, i, is_neg) = @inbounds NTuple{3,UInt8}((0x32, 0x35, 0x35))[i]
 _typemaxbytes(::Type{UInt16}, i, is_neg) = @inbounds NTuple{5,UInt8}((0x36, 0x35, 0x35, 0x33, 0x35))[i]
 _typemaxbytes(::Type{UInt32}, i, is_neg) = @inbounds NTuple{10,UInt8}((0x34, 0x32, 0x39, 0x34, 0x39, 0x36, 0x37, 0x32, 0x39, 0x35))[i]
-_typemaxbytes(::Type{UInt64}, i, is_neg) = @inbounds NTuple{19,UInt8}((0x31, 0x38, 0x34, 0x34, 0x36, 0x37, 0x34, 0x34, 0x30, 0x37, 0x33, 0x37, 0x30, 0x39, 0x35, 0x35, 0x31, 0x36, 0x31, 0x35))[i]
+_typemaxbytes(::Type{UInt64}, i, is_neg) = @inbounds NTuple{20,UInt8}((0x31, 0x38, 0x34, 0x34, 0x36, 0x37, 0x34, 0x34, 0x30, 0x37, 0x33, 0x37, 0x30, 0x39, 0x35, 0x35, 0x31, 0x36, 0x31, 0x35))[i]
```
[the remark this answers is not in the document]
Comment (NHDaly) [reply src/fixed_decimals_utils.jl:134]: Are these corresponding to strings? Like `"127"` for the first one? Ah yeah i think they are:
```julia
julia> codeunits("127")
3-element Base.CodeUnits{UInt8, String}:
 0x31
 0x32
 0x37
```

------

If so, can you build these that way? Like, change these lines to:
```julia
_typemaxbytes(::Type{Int8}, i, is_neg) = @inbounds Tuple(codeunits("127"))[i] + ((i == 3) * is_neg)
_typemaxbytes(::Type{Int16}, i, is_neg) = @inbounds Tuple(codeunits("32767"))[i] + ((i == 5) * is_neg)
```
or maybe:
```julia
_bytes(::Type{Int8}) = Tuple(codeunits("127"))
@eval _typemaxbytes(::Type{Int8}, i, is_neg) = @inbounds $(_bytes(Int8))[i] + ((i == 3) * is_neg)
@eval _typemaxbytes(::Type{Int16}, i, is_neg) = @inbounds $(_bytes(Int16))[i] + ((i == 5) * is_neg)
```
or even:
```julia
for T in Base.BitSigned_types
    bytes = Tuple(codeunits(string(typemax(T))))
    @eval _typemaxbytes(::Type{$T}, i, is_neg) = @inbounds $bytes[i] + ((i == $(length(bytes))) * is_neg)
end
for T in Base.BitUnsigned_types
    bytes = Tuple(codeunits(string(typemax(T))))
    @eval _typemaxbytes(::Type{$T}, i, is_neg) = @inbounds $bytes[i] + ((i == $(length(bytes))) * is_neg)
end
```

..... or another approach is to keep the code how you have it, so that it's staying explicit, but to have **tests** that test that each of the functions match the expected value, like:
```julia
for T in Base.BitSigned_types
    bytes = Tuple(codeunits(string(typemax(T))))
    for (i,b) in enumerate(bytes)
        @test _typemaxbytes(T, i, false) == bytes[i]
    end
    @test _typemaxbytes(T, i, false) == bytes[i] + 1
end
...
``` 
or something like that?
Or vice versa: do the generated thing here and test against the hardcoded values?

[the remark this answers is not in the document]
Comment (NHDaly) [reply src/fixed_decimals_utils.jl:134]: I think at the very least, there ought to be a comment with a couple examples, explaining what this means and why it's here. You should really have that for any block of unexplained values like this.

-------

The goal with my suggestions here would be to _prevent_ exactly the kind of bug you ran into here, by automating this instead of hardcoding it.

(😁 this situation even came up in the [article you linked me](http://number-none.com/blow/john_carmack_on_inlined_code.html)!)

> On the subject of code duplication, I tracked all the bugs (problems that surfaced after I thought everything was working correctly) I fixed for a while, and I was quite surprised at how often copy-paste-modify operations resulted in subtle bugs that weren't immediately obvious. For small vector operations on three or four things, I would often just past and modify a few characters like this:
> 
>  
> ```
> v[0] = HF_MANTISSA(*(halfFloat_t *)((byte *)data + i*bytePitch+j*8+0));
> v[1] = HF_MANTISSA(*(halfFloat_t *)((byte *)data + i*bytePitch+j*8+1));
> v[2] = HF_MANTISSA(*(halfFloat_t *)((byte *)data + i*bytePitch+j*8+2));
> v[3] = HF_MANTISSA(*(halfFloat_t *)((byte *)data + i*bytePitch+j*8+3));
> ```
> 
> I now strongly encourage explicit loops for everything, and hope the compiler unrolls it properly. A significant number of my bugs related to things like this, and I am now rethinking even two dimensional cases where I commonly use discrete _X, _Y or _WIDTH, _HEIGHT variables. I find that much easier to read than a two element array, but it is difficult to argue with my data on how often it made me screw up.
>

[the remark this answers is not in the document]
Comment (Drvi) [reply src/fixed_decimals_utils.jl]: I've replaced this code with a bit of metaprogramming :+1: All of this is tested, that's how I discovered the bug. I've also improved the comment.

> I can't tell from reading this whether 19 or 20 is the right number, and i find myself just trusting your change on faith

In this case, one way to do it was to count the number of elements in that tuple or to observe that the original version errors while the new one doesn't.

File: src/fixed_decimals_utils.jl [hunk 9]
```diff
@@ -1,410 +1,508 @@
...
-    elseif decpos ==  2 && return T(100) * n
-    elseif decpos ==  3 && return T(1000) * n
+    if     decpos == 0 && return n
+    elseif decpos == 1 && return T(10) * n
+    elseif decpos == 2 && return T(100) * n
+    elseif decpos == 3 && return T(1000) * n
     else
         @assert false "unreachable"
     end
 end
```
Comment (NHDaly) [inline src/fixed_decimals_utils.jl:111]: For next time, it would be good to separate the whitespace changes from the functional changes.

I missed that this function was adding a `== 0` case until i disabled whitespace in the GitHub settings. I think this is good practice in general, and it was a requirement at Google.

Basically, the main principle to internalize is to prioritize the reviewer rather than the author whenever sending code for review. If it's easier to review split into many smaller PRs, that's better, if it's easier to review by making a copy of the files rather than changing them, then following up by deleting the old code in a separate PR, that's better. Basically that's the main principle and the details can change based on your judgement.

File: src/fixed_decimals_utils.jl [hunk 10]
```diff
@@ -1,410 +1,508 @@
...
-                break
+# `_typeextremabytes` is used to recognize byte patterns corresponding to typemin and typemax
+# of supported integer types, e.g. 0x31, 032, 0x37 for b"127", the typemax of Int8,
+# or 0x31, 032, 0x38 for b"(-)128", the typemin of Int8. This is used for overflow detection.
+# `maxdigits` tells us the number of digits the type can hold, e.g. 3 for Int8 (for 127)
+for T in Base.BitSigned_types
+    let bytes = Tuple(codeunits(string(typemax(T))))
+        @eval _typeextremabytes(::Type{$T}, i, is_neg::Bool) = @inbounds $(bytes)[i] + ((i == $(length(bytes))) * is_neg)
+        @eval maxdigits(::Type{$T}) = $(length(bytes))
+    end
```
Comment (NHDaly) [inline src/fixed_decimals_utils.jl:122]: I think that this is unsafe to call `@inbounds` on an argument that the callee doesn't control.

If someone called `_typeextremabytes(Int8, 6)` this could segfault, give wrong results, etc. 

This kind of bug waiting to happen from `@inbounds` on arguments the function didn't independently validate is one of the issues that Yuri was pointing out in his blog post:
https://discourse.julialang.org/t/discussion-on-why-i-no-longer-recommend-julia-by-yuri-vishnevsky/81151

I think the safe pattern is to first include a boundscheck and mark that with `@boundscheck`, and then mark the function as `@propagate_inbounds`, and finally to put `@inbounds` on the *call site*, since its the caller's responsibility to ensure that they're passing in an `i` that actually **is** in bounds.

Comment (NHDaly) [reply src/fixed_decimals_utils.jl:122]: Of course since it's an internal private function, currently it's safe since you know you're always calling it with the right value. But this is a future-maintenance bug in waiting, since implementations change over time

File: test/decimals.jl [hunk 11]
```diff
@@ -1,225 +1,396 @@
...
+                test_decimal_parser(T, 0, "+" * test_case, expect=123, options=test_options)
+                if typemin(T) < 0
+                    test_decimal_parser(T, 0, "-" * test_case, expect=-123, options=test_options)
+                end
+            end
+        end
+    end
+end
+
+end # @testset "decimals"
```
Comment (NHDaly) [inline test/decimals.jl:396]: V glad to see the improved unit tests!! 💪 :)

File: src/fixed_decimals_utils.jl [hunk 12]
```diff
@@ -1,410 +1,508 @@
...
+    found_digit::Bool
+    # The position of the first byte to parse
+    first_byte_to_parse::Int
+    # The position of the last byte to parse
+    last_byte_to_parse::Int
+end
+
+@inline function _parse_digits_after_scanning(
+    ::Type{T}, f, source, pos, len, b, code, options, mode, ctx::DecimalParsingContext
+) where {T}
```
Comment (NHDaly) [inline src/fixed_decimals_utils.jl:157]: For symmetry, and ease of understanding, can you break out a function for the first pass over the data, call it something like `_scan_for_digits()`, which returns the `DecimalParsingContext` and put that before this function?

Then the implementation of `_typeparser()` could be simplified to first calling the first function and then subsequently calling the second function.

Something like
```
_typeparser(...)
    ctx_or_nothing, failure =  _scan_for_digits(...)
    if failure !== nothing
        failure isa EOF && code |= Parsers.EOF
        return (out, code |= Parsers.INVALID, pos, inexact)
    end
    return _parse_digits_after_scanning(T, f, source, pos, len, b, code, options, mode, ctx)
end
```
or something like that?

File: src/fixed_decimals_utils.jl [hunk 13]
```diff
@@ -1,410 +1,508 @@
...
+# digits and calculating the precision and second one to parse the number and round it.
+# It should be possible to do a single-pass implementation, but it's not trivial.
+"""
+    _typeparser(
+        ::Type{T},
+        f::Int,
+        buf::Union{AbstractVector{UInt8},AbstractString,IO},
+        pos::Int,
+        len::Int,
+        b::UInt8,
```
Comment (NHDaly) [inline src/fixed_decimals_utils.jl:283]: what is `b`? is it the first byte? can it be called `first_byte`?

File: src/fixed_decimals_utils.jl [hunk 14]
```diff
@@ -1,410 +1,508 @@
...
+`RoundDown` rounds towards zero).
+"""
 Base.@propagate_inbounds _typeparser(::Type{T}, f, buf::AbstractString, pos, len, b, code, options::Parsers.Options, mode::Base.RoundingMode=Base.RoundNearest) where {T<:Union{Int8,Int16,Int32,Int64,Int128,UInt8,UInt16,UInt32,UInt64,UInt128}} =
     _typeparser(T, f, codeunits(buf), pos, len, b, code, options, mode)
-Base.@propagate_inbounds function _typeparser(::Type{T}, f, buf::AbstractVector{UInt8}, pos, len, b, code, options::Parsers.Options, mode::Base.RoundingMode=Base.RoundNearest) where {T<:Union{Int8,Int16,Int32,Int64,Int128,UInt8,UInt16,UInt32,UInt64,UInt128}}
-    decimal = options.decimal
+Base.@propagate_inbounds function _typeparser(::Type{T}, f, source, pos, len, b, code, options, mode::Base.RoundingMode=Base.RoundNearest) where {T}
+    out = zero(T)
+    has_groupmark = !isnothing(options.groupmark)
+    groupmark = something(options.groupmark, 0xff)
```
Comment (NHDaly) [inline src/fixed_decimals_utils.jl:303]:
```
julia> Char(0xff)
'ÿ': Unicode U+00FF (category Ll: Letter, lowercase)
```
Is this right? If you don't set a groupmark option, it defaults to this thingy? 😅  should we even _allow_ you to not set a groupmark? Why support that option?

File: src/fixed_decimals_utils.jl [hunk 15]
```diff
@@ -1,410 +1,508 @@
...
-    end
+    last_byte_to_parse = 0
+    first_byte_to_parse = Int(pos + (b == options.decimal)) # will be adjusted later if we have leading zeros
 
-    # Remove leading zeros in 0.00..x
-    if field_end > pos && buf[pos] == UInt8('0')
-        if field_end > 1 + pos
-            if buf[1+pos] == decimal
+    # The digits before the decimal point / exponent
+    prev_b = groupmark + 0x01
```
Comment (NHDaly) [inline src/fixed_decimals_utils.jl:328]: can you explain this line? it's nonobvious to me

File: src/fixed_decimals_utils.jl [hunk 16]
```diff
@@ -1,410 +1,508 @@
...
+                !found_nonzero && (first_byte_to_parse = pos)
+                found_nonzero = true
+                last_nonzero_digit_pos = pos
+                pos += 1
+                Parsers.incr!(source)
+                num_digits += 1
+                Parsers.eof(source, pos, len) && (code |= Parsers.EOF; @goto done_scanning) # Input is something like "11<EOF>"
+            elseif has_groupmark && b == groupmark
+                # More than one consecutive groupmark is invalid
+                prev_b == groupmark && @goto invalid # Input is something like "10,,"
```
Comment (NHDaly) [inline src/fixed_decimals_utils.jl:358]: Minor: it's probably not meaningful, but i guess you could save one iteration on invalid inputs by checking the peek lookahead for consecutive groupmarks instead of checking the prev_b?

probably too small to notice in reality tho.

Comment (NHDaly) [reply src/fixed_decimals_utils.jl:358]: oh, you do that right on the next couple lines.So you could just add the check to right before 361?

Comment (NHDaly) [reply src/fixed_decimals_utils.jl:358]: ehhh in retrospect i kind of like the symmetry the current code has with the decimal after groupmark check on 374. So i'm more meh about my suggestion

File: src/fixed_decimals_utils.jl [hunk 17]
```diff
@@ -1,410 +1,508 @@
...
-
-    decimal_shift = f - number_of_fractional_digits
-
-    backing_integer_digits = number_of_digits - number_of_fractional_digits + f - ngroupmarks
-    digits_to_iter = min(backing_integer_digits, number_of_digits)
+    # The fractional digits
+    if b == options.decimal
+        prev_b == groupmark && @goto invalid # Input is something like "10,."
+        decimal_position = pos
+        pos += 1
```
Comment (NHDaly) [inline src/fixed_decimals_utils.jl:376]: Shouldn't we also check that there isn't more than one decimal point?

Something like 
```julia
        decimal_position != 0 && @goto invalid # Multiple decimal points aren't allowed
```

if that's right, can you also add tests for this?

File: src/fixed_decimals_utils.jl [hunk 18]
```diff
@@ -1,410 +1,508 @@
...
-        end
-        out *= int_sign
-        if number_of_fractional_digits < f
-            out = _shift(out, decimal_shift)
-        elseif number_of_fractional_digits != f
-            round_val, code = _parse_round(T, buf, pos + digits_to_iter + parse_through_decimal + ngroupmarks, last_byte_to_parse, decimal_position, code, mode)
-            out += int_sign * round_val
+        # If we started on a decimal point or with leading zeros only, we'll keep skipping zeros
+        if !found_nonzero
+            found_digit |= b == UInt8('0')
```
Comment (NHDaly) [inline src/fixed_decimals_utils.jl:383]: This might be easier to read with parens:
```suggestion
            found_digit |= (b == UInt8('0'))
```

**Review by NHDaly — COMMENTED** [review]
I've been enjoying reading this! Sorry that i'm slow. I'm having to come at it in spurts when i get some time.

I'm almost done with the first-pass function. 👍 I agree that the new code is definitely more readable than the old code you're replacing. Thanks for that!

Will finish soon.

File: src/fixed_decimals_utils.jl (hunk 17 above)
Comment (Drvi) [reply src/fixed_decimals_utils.jl:376]: Parsers.jl work like this -- you try to parse as many bytes for your type as possible until you encounter an invalid byte. Let's say your position is `*`. Then `10.0e*.` is clearly an invalid situation and you as a type-parser need to say "this is invalid, after an `e`, there must be a digit or a -/+, not this" and mark the code as `INVALID` and yield back to Parsers.jl who will jump to next delimiter. In a different example, `10.0*.`, you are at a place where you can end with type-parsing, you already have a safe value to report as `10.0` is valid number, so you yield back to Parsers.jl to finish stuff up. Parsers.jl will notice that the next byte is a `.` which is not a valid thing to skip nor a valid delimiter, so it will mark the status code as invalid (with `INVALID_DELIMITER`).

So I don't think we need any special handling for this, but I can add a testcase.

File: src/fixed_decimals_utils.jl (hunk 15 above)
Comment (Drvi) [reply src/fixed_decimals_utils.jl:328]: Later in fractional / exponent parsing, we check if the last byte before `.`/`e` was a `groupmark` so we can check for invalid things like `10,.` or `10,e`. So here we initialize `prev_b` to be anything else than a `groupmark` in case there was no integer part.

I think we currently do allow ending on a `groupmark` `10,<EOF>`, not sure if want to be stricter about that?

File: src/fixed_decimals_utils.jl (hunk 14 above)
Comment (Drvi) [reply src/fixed_decimals_utils.jl:303]: `nothing` for groupmark means there are no group marks allowed in the thing I want to parse, and since we guard `groupmark` skipping with `has_groupmark` I just chose garbage value here.

File: src/fixed_decimals_utils.jl (hunk 17 above)
Comment (NHDaly) [reply src/fixed_decimals_utils.jl:376]: Aha, very cool! Yeah that was not obvious to me from reading the code; I guess it's something you just need to know about how Parsers works.

Yes, test cases would be good, and maybe a tiny comment around where I was asking?

File: src/fixed_decimals_utils.jl (hunk 14 above)
Comment (NHDaly) [reply src/fixed_decimals_utils.jl:303]: gotcha. That needs a comment. Not clear from reading.

Also, would it make sense to change it to only assign it if `has_groupmark` is true, since you're not going to use it otherwise?

File: src/fixed_decimals_utils.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (NHDaly) [reply src/fixed_decimals_utils.jl]: > In this case, one way to do it was to count the number of elements in that tuple or to observe that the original version errors while the new one doesn't.

Once I read more carefully and understood it, yeah that's what I did. But it was hard for me to even know what these meant at first, which is why i was confused, and which is why more comments and/or a restructuring were needed.

👍 the metaprogramming version looks good. It's clearer what it is doing and it's less error prone 👍

**Comment by Drvi** [comment]
@NHDaly Thank you for your review :+1: Just FYI @quinnj gave me an idea to reuse some floating point parsing code from Parsers.jl, so I'm looking into that. I think it would be best to wait until I'm done with that as it should save a lot of code.
